internal/auth/internal/controllers/trafficpermissions/controller.go

```diff
@@ -0,0 +1,181 @@
+
+	// GetTrafficPermissionsForCTP returns the tracked TrafficPermissions that are used to create a CTP
+	GetTrafficPermissionsForCTP(id *pbresource.ID) []*pbresource.Reference
+}
+
+// Controller creates a controller for automatic ComputedTrafficPermissions management for
+// updates to WorkloadIdentity or TrafficPermission resources.
+func Controller(mapper ComputedTrafficPermissionsMapper) controller.Controller {
+	if mapper == nil {
+		panic("No ComputedTrafficPermissionsMapper was provided to the TrafficPermissionsController constructor")
+	}
+
+	return controller.ForType(types.ComputedTrafficPermissionsType).
+		WithWatch(types.WorkloadIdentityType, controller.ReplaceType(types.ComputedTrafficPermissionsType)).
+		WithWatch(types.TrafficPermissionsType, mapper.MapTrafficPermissions).
+		WithReconciler(&reconciler{mapper: mapper})
+}
+
+type reconciler struct {
+	mapper ComputedTrafficPermissionsMapper
+}
+
+// Reconcile will reconcile one ComputedTrafficPermission (CTP) in response to some event.
+// Events include adding, modifying or deleting a WorkloadIdentity or TrafficPermission.
+func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
+	rt.Logger = rt.Logger.With("resource-id", req.ID, "controller", StatusKey)
+	/*
+	 * A CTP ID could come in for a variety of reasons.
+	 * 1. workload identity create / delete: this results in the creation / deletion of a new CTP
+	 * 2. traffic permission create / modify / delete: this results in a potential modification of an existing CTP
+	 *
+	 * Part 1: Handle Workload Identity changes:
+	 * Check if the workload identity exists. If it doesn't we can stop here.
+	 * CTPs are always generated from WorkloadIdentities, therefore the WI resource must already exist.
+	 * If it is missing, that means it was deleted.
+	 */
+	ctpID := req.ID
+	wi := &pbresource.ID{
+		Type:    types.WorkloadIdentityType,
+		Tenancy: ctpID.Tenancy,
+		Name:    ctpID.Name,
+	}
+	workloadIdentity, err := resource.GetDecodedResource[*pbauth.WorkloadIdentity](ctx, rt.Client, wi)
+	if err != nil {
+		rt.Logger.Error("error retrieving corresponding Workload Identity", "error", err)
+		return err
+	}
+	if workloadIdentity == nil || workloadIdentity.Resource == nil {
+		rt.Logger.Trace("workload identity has been deleted")
+		// The workload identity was deleted, so we need to update the mapper to tell it to
+		// stop tracking this workload identity, and clean up the associated CTP
+		r.mapper.UntrackWorkloadIdentity(ctpID)
+		return nil
+	}
+
+	// Check if CTP exists:
+	oldCTPData, err := resource.GetDecodedResource[*pbauth.ComputedTrafficPermissions](ctx, rt.Client, ctpID)
+	if err != nil {
+		rt.Logger.Error("error retrieving computed permissions", "error", err)
+		return err
+	}
+	if oldCTPData == nil {
+		// CTP does not yet exist, so we need to make a new one
+		rt.Logger.Trace("creating new computed traffic permissions for new workload identity")
```

Comment:
how could it not be? all workload identities have a CTP created when they first get reconciled. if the workload identity exists in the resource service and does not have a CTP, this must be the first time we have seen it.

Reply:
It wouldn't be new if a user deleted the CTP without deleting workload identity, it won't be new.

Reply:
If we want to account for direct user writes to CTPs, then we should probably also watch CTP resources and reconcile changes to them. There aren't test cases for this right now, but I can add them.

Reply:
We are already watching them with:

```
controller.ForType(types.ComputedTrafficPermissionsType)
```

>There aren't test cases for this right now, but I can add them.

We should definitely add test cases for it. 

>We should consider why this would happen, because the reconciler will simply revert to the state dictated by traffic permisisons and workload identities.

Generally, for computed resources this shouldn't happen, however, if a user needs break-glass access they should be able to do it. For example, something is wrong with their cluster and they want to delete CTP and have the controller re-compute it again.

Reply:
User deletion of CTPs raises many questions about when and how the state can get fully re-reconciled.

Reply:
Do we even allow direct CRUD operations on the CTP? My understanding was the user can only directly interact with `TrafficPermissions` and `WorkloadIdentities` and that the CTP are purely synthetic resources we create and manage. 

To force a recompute, you could just change any of the traffic permissions I think.